components/validator/guardian.go

```diff
@@ -103,121 +91,101 @@ func (g *Guardian) ValidateL2Output(ctx context.Context, outputRoot eth.Bytes32,
-	cCancel()
+	txOpts := utils.NewSimpleTxOpts(ctx, g.cfg.TxManager.From(), g.cfg.TxManager.Signer)
+	return g.securityCouncilContract.ConfirmTransaction(txOpts, transactionId)
+}
+
+func (g *Guardian) handleValidationRequested(ctx context.Context) {
+	defer g.wg.Done()
+	for {
+		select {
+		case ev := <-g.validationRequestedChan:
+			g.wg.Add(1)
+			go g.processOutputValidation(ctx, ev)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (g *Guardian) tryConfirmTransaction(ctx context.Context, event *bindings.SecurityCouncilValidationRequested) error {
+	needConfirm, err := g.checkConfirmCondition(ctx, event.TransactionId)
 	if err != nil {
-		return nil, err
+		// error occurred. go to retry.
+		return err
 	}
-	if executionTx.Executed {
-		log.Warn("transaction %d already executed.", transactionId)
-		return nil, nil
+	if !needConfirm {
+		return nil
 	}
 
-	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
-	executionGas, err := g.cfg.L1Client.EstimateGas(cCtx, ethereum.CallMsg{
-		From:  g.cfg.SecurityCouncilAddr,
-		To:    &executionTx.Destination,
-		Value: executionTx.Value,
-		Data:  executionTx.Data,
-	})
-	cCancel()
+	isValid, err := g.ValidateL2Output(ctx, event.OutputRoot, event.L2BlockNumber.Uint64())
 	if err != nil {
-		return nil, fmt.Errorf("failed to estimate confirmation tx gas: %w", err)
+		return fmt.Errorf("failed to verify the output to be replaced. (transactionId: %d): %w", event.TransactionId.Int64(), err)
 	}
+	if isValid {
+		g.log.Info("the output to be replaced is valid")
 
-	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
-	txOpts := utils.NewSimpleTxOpts(cCtx, g.cfg.TxManager.From(), g.cfg.TxManager.Signer)
-	confirmTx, err := g.securityCouncilContract.ConfirmTransaction(txOpts, transactionId)
+		cCtx, cCancel := context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+		tx, err := g.ConfirmTransaction(cCtx, event.TransactionId)
+		cCancel()
+		if err != nil {
+			return fmt.Errorf("failed to create confirm tx. (transactionId: %d): %w", event.TransactionId.Int64(), err)
+		}
+		if tx == nil {
+			return nil
+		}
+
+		if txResponse := g.cfg.TxManager.SendTransaction(ctx, tx); txResponse.Err != nil {
+			return fmt.Errorf("failed to send confirm tx. (transactionId: %d): %w", event.TransactionId.Int64(), txResponse.Err)
+		}
+	} else {
+		g.log.Info("ignore this challenge because the output to be replaced is invalid")
+	}
+	return nil
+}
+
+func (g *Guardian) checkConfirmCondition(ctx context.Context, transactionId *big.Int) (bool, error) {
+	cCtx, cCancel := context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	callOpts := utils.NewCallOptsWithSender(cCtx, g.cfg.TxManager.From())
+	isConfirmed, err := g.securityCouncilContract.IsConfirmed(callOpts, transactionId)
 	cCancel()
 	if err != nil {
-		return nil, err
+		return true, fmt.Errorf("failed to get confirmation. (transactionId: %d): %w", transactionId.Int64(), err)
```

Comment:
I think if failed to get confirmation, we don't know if confirmation is needed or not. Then the default value of `needConfirm` is better to be `false`. wdyt?
```suggestion
		return false, fmt.Errorf("failed to get confirmation. (transactionId: %d): %w", transactionId.Int64(), err)
```